### ISSUE TITLE
Orthologs with negative prediction scores

### ISSUE THREAD

**Issue by alejandrogzi** [hillerlab/TOGA#70]
Hi @MichaelHiller and @kirilenkobm,

While inspecting TOGA results I found that some of the transcripts annotated had negative prediction scores:

`ENST00000373688	209092	-1.0` 

This, in theory, would not be possible since XGBoost outputs a probability. Surprisingly, that gene has been annotated in the final .bed file:

`chr21	32269051	32269996	ENST00000373688.209092	1000	-	32269051	32269996	0,0,200	1	945,	0,`

The same scenario occurs for 12 other transcripts. Could you clarify this situation, please?

Thanks,
Alejandro

**Comment by MichaelHiller** [comment]
Interesting. We assign -1 as the chain ID in case TOGA joins several orthologous gene fragments, but I think this is different from what you describe. 

From which file is the ENST00000373688	209092	-1.0 line? 

Can you pls have a look whether the chain 209092 actually aligns exons of the gene? 
If that is not the case, the chain could span the gene (align stuff up and downstream) and we can't run the classifier as the chain would not have features for the focal gene. But if the span is not too large, we still run CESAR, which may predict a transcript. 

@kirilenkobm Can this be the case? We need to document this.

**Comment by alejandrogzi** [comment]
@MichaelHiller,

Thanks for your quick response. 

> From which file is the ENST00000373688 209092 -1.0 line?

That information is coming from the orthology_score.tsv file. Plotting all orthology scores and comparing them with their Ensembl counterparts (as you did in the TOGA paper) revealed all those rare transcripts annotated with negative orthology scores.

That chain is coming from:

`chain 36727 chr9 138394717 + 122510811 122568396 chr21 52140716 - 19802762 19880901 209092`

and does not align any exon of that particular transcript:

[ENST00000373688.pdf](https://github.com/hillerlab/TOGA/files/11711770/ENST00000373688.pdf)

I also was curious about what does the metadata of that exon says:

`ENST00000373688	0	209092	chr21:32269996-32269051	exp:32260659-32290392	INC	52.03	40.18	OK	C	False	LQ` (exons_meta_data.tsv)
`ENSG00000171501	ENST00000373688	reg_8141	ENST00000373688.209092	many2many` (orthology_classification.tsv)
`PROJECTION	ENST00000373688.209092	I` (loss_summ_data.tsv)
`ENST00000373688.209092	low_confidence` (transcript_quality.tsv)

I'll check the chains of the other problematic transcripts to confirm the same behavior. Please let me known if you need more information.

Alejandro

**Comment by MichaelHiller** [comment]
Thanks for sending the screenshot. Yes, that chain spans the single exon ENST00000373688 (therefore we cannot run the classifier and output prob = -1 to indicate that) and CESAR finds an alignment. In general, these are more tricky cases with 2 possible scenarios. 
1) The gene could be highly diverged (but still orthologous), thus it does not align at the nucleotide level (more likely if there is only a single exon), hence the chain spans the gene. But CESAR (coding to nucleotide alignment) may detect the gene. 
2) No ortholog might exist in this spanned locus in the query (deletion, or rearrangement), but CESAR finds a bogus alignment that passes our filters (I think Fig S12 in the paper). Such edge cases likely also exist. 

If you care about these ~10 cases, pls check if the dog transcriptomes (there are ESTs/mRNAs and also RNA-seq) support the predicted exons.

**Comment by kirilenkobm** [comment]
@alejandrogzi @MichaelHiller 

Sorry for the confusion.
-1 in the query_annotation.bed means that the transcript was joined from fragments.
-1 in the orthology scores file means that the chain is spanning - has no alignment to the coding part of the gene. For such cases, we cannot compute the full set of features properly, and xgboost model is inapplicable.
In 99% of cases, it means that the gene in the respective locus is either missing or is deleted.
However, TOGA still tries to annotate such cases.

This is something to be documented in wiki: https://github.com/hillerlab/TOGA/wiki/Understantding-orthology-scores
the page is quite drafty for now.

**Comment by alejandrogzi** [comment]
@MichaelHiller @kirilenkobm,

I did not want to open a new issue because this is a interpretation-related question rather than something linked with TOGA itself.

Some context: I am trying to build a 1:1 high-confidence ortholog annotation (to apply something @MichaelHiller did with cross-species transcriptomics in jerboas in 2022). To achieve this, I am not only considering TOGA as my source but trying to add all Ensembl 1:1 orthologs. As expected the overlap is high, but I am also giving attention to TOGA-only and Ensembl-only 1:1 orthologs. TOGA detects most of the 1:1 Ensembl-only; however, classify them in another orthology class different than 1:1. Since some of those are high-confidence in the Ensembl database, I wanted to inspect them manually.

This is one of the cases where seems that TOGA misclassify a 1:1 relationship into 1:many class: 

[PTEN_case.pdf](https://github.com/hillerlab/TOGA/files/12098884/PTEN_case.pdf)

This case is important to consider in a gene annotation, since _PTEN_ is a tumor-suppressor gene (TSG) conserved in different species (and particularly with the species I am working with).





t_gene | t_transcript | q_gene | q_transcript | orthology_class
-- | -- | -- | -- | --
ENSG00000171862 | ENST00000700022 | reg_17643 | ENST00000700022.138 | one2many
ENSG00000171862 | ENST00000693560 | reg_17643 | ENST00000693560.138 | one2many
ENSG00000171862 | ENST00000472832 | reg_17643 | ENST00000472832.138 | one2many
ENSG00000171862 | ENST00000700021 | reg_17643 | ENST00000700021.138 | one2many
ENSG00000171862 | ENST00000706955 | reg_17643 | ENST00000706955.138 | one2many
ENSG00000171862 | ENST00000706954 | reg_17643 | ENST00000706954.138 | one2many
ENSG00000171862 | ENST00000686459 | reg_17643 | ENST00000686459.138 | one2many
ENSG00000171862 | ENST00000706953 | reg_17712 | ENST00000706953.138 | one2many
ENSG00000171862 | ENST00000688922 | reg_17643 | ENST00000688922.138 | one2many
ENSG00000171862 | ENST00000688308 | reg_17643 | ENST00000688308.138 | one2many
ENSG00000171862 | ENST00000371953 | reg_17643 | ENST00000371953.138 | one2many

Could you please give some advice in this case or explanation why TOGA classify this as one2many when clearly all projections are coming from the same alignment chain? Also some insights about using TOGA to generate a 1:1 high-confidence annotation? 

Thanks in advance,
Alejandro

**Comment by MichaelHiller** [comment]
Hi Alejandro, 

thanks for posting this. I agree, based on the single chain, it should be a 1:1 ortholog. 
For the many mammals we have TOGA runs, it is also a clear 1:1 ortholog, but we include only ENST00000371953 in our transcript set (the other transcripts were filtered out).

In your case, TOGA seems to assign the different PTEN transcripts (all belonging to the ENSG00000171862 gene) to different query genes that exist in the same locus. 
My guess is that you did not specify the isoform file that tells TOGA that the different transcripts all belong to the same file. Can you pls check that? 

Otherwise, Bogdan likely has an idea. 

Thx

**Comment by alejandrogzi** [comment]
@MichaelHiller,

Thanks for your quick response. I specified the isoforms file with care that all transcripts occurring in the bed file appear there too.

This is the file that I used:
[isoforms.txt](https://github.com/hillerlab/TOGA/files/12100423/isoforms.txt)

Just to give more context, I am working with human/dog chains using the ROS_Cfam assembly.

**Comment by MichaelHiller** [comment]
OK, did you also specify the 
--isoforms, -i ISOFORMS_FILE
argument to tell TOGA to use that file? 

Can you send me also your transcripts.bed file, only with the PTEN transcripts? I can try to recap this behaviour here.

**Comment by alejandrogzi** [comment]
This was the command that I used:

`./toga.py human.dog.allfilled.chain.gz hg38.GENCODE.v43.22X.bed human.2bit dog.2bit --isoforms isoforms.txt --u12 hg38.U12sites.tsv --keep_temp --project_name TOGA_GRCh38_ROSCfam --cesar_jobs_num 500 --cesar_buckets 3,10,30,60,100,300 --ms --nd ~/nextflow_logs --nc ~/nextflow_config_files --do_not_del_nf_logs`

hg38.GENCODE.v43.22X.bed with only PTEN transcripts: https://drive.google.com/file/d/1BTYn9fnGzYq__6Ue4o3aRzjzYE_1FWvS/view?usp=sharing

Thanks for your time, @MichaelHiller

**Comment by MichaelHiller** [comment]
I figured it out. When I take the transcripts you provide, I also get 1:many running it against canFam5. 

First, you have a few transcripts that are non-coding (those have CDSstart == CDSend). I removed these transcripts with 
  cat toga.transcripts.includingNoncoding.bed | awk '{if ($7-$8 <0) print $0}' > toga.transcripts.bed 
@kirilenkobm  Maybe we can add a filter to remove non-coding transcripts at the very beginning of TOGA and log which transcripts are excluded because they are non-coding. 

Second, the issue is ENST00000706953. This is a single exon transcript without a UTR that does not overlap any other PTEN transcripts. In dog, this projected transcript also does overlap the other PTEN transcripts. Since TOGA uses "same strand exon overlap" to assign projected transcripts to query genes and since ENST00000706953 does not overlap the other PTENs, this looks to TOGA like 2 gene loci in the query, hence 1:2. 
--> Either remove that transcript (not sure if this is real, has no UTR ...). Or assign that transcript to another gene (say PTEN_like). Then you get two 1:1 orthologs.

![image](https://github.com/hillerlab/TOGA/assets/8644098/5e0bbc9c-e846-4134-8d15-b06d6b1d0b18)

**Comment by alejandrogzi** [comment]
Brilliant! 

Thanks for your help, Michael! I will implement those filters. This is one particular case of ~2000 transcripts with inconsistent orthology relationships between TOGA and Ensembl, so I will take care of all non-coding transcripts + transcripts with no UTR's (supposing all of them will probably generate similar problems). I will also evaluate the possibility of transcript-like assignment.

I will let you know the results!

Best,
Alejandro

**Comment by alejandrogzi** [comment]
Hi again @MichaelHiller & @kirilenkobm,

Results were great and I captured previously non-1-to-1 relationships that were clearly 1-to-1. I have one more question:

here: 

> we include only ENST00000371953 in our transcript set (the other transcripts were filtered out).

you wrote that in the PTEN case the rest of transcripts were filtered out. I have encountered these other problem were a huge amount of isoforms are annotated for a gene. This, obviously, is due to the annotation provided to TOGA. Would you please share what was your approach to filter transcripts? (maybe the longest isoform?).


Thanks,
Alejandro

**Comment by MichaelHiller** [comment]
We provide the filtered transcripts for hg38, mm10 and galGal6 here on github: https://github.com/hillerlab/TOGA/tree/master/TOGAInput/. 
Longest transcript is often not a good idea, as these may have lineage-specific exons or exon extensions (regions that are non-coding in other lineages). The way we filtered the transcripts is described  in the methods section, copied here: 

"First, all
noncoding transcripts that lack an annotated
CDS are excluded. Second, we excluded isoforms
with a CDS that is too short. We compute
for each gene a CDS length threshold as
80% of the CDS length of the principal APPRIS
isoform. If the gene has more than one principal
isoform, we used the principal isoform
with the shortest CDS. If APPRIS does not
provide a principal isoform for the gene, we used the transcript with the longest CDS instead.
We then excluded all transcripts that
have CDS length below this threshold. Third,
we excluded erroneous transcripts that have a
CDS length not divisible by 3. Fourth, we excluded
potential NMD targets that have the
annotated stop codon more than 55 bp upstream
of the last exon-exon junction. Fifth, we
excluded isoforms that have introns shorter
than 20 bp, because such micro-introns are
often used to mask frameshifting mutations.
Sixth, if several isoforms have an identical
coding region, we selectedonly the onewith the
longest UTR. This step reduces redundancy
because TOGA only annotates the CDS. Seventh,
we excluded transcripts that have in-frame stop
codons unless the stop codon(s) is a TGA codon,
in which case it may encode selenocysteine.
Finally,we excluded transcripts that do not start
with an ATG codon or end with a stop codon.
For genes with many transcripts, these filters
ensure that only proper transcripts will be
used as input for TOGA.However, it is possible
that these filters eliminate all transcripts of a
gene, for example, if the reference genome
has a base error in a constitutive exon. Because
this would result in missing the gene
entirely, we include for such genes the longest
transcript that has a CDS length divisible
by three."
